src/seccomp_utils.c

```diff
@@ -22,25 +22,393 @@
 #include <stdio.h>
 #include <unistd.h>
 #include <string.h>
+#include <search.h>
 #include <ctype.h>
 #include <stdbool.h>
 #include <stdlib.h>
+
+// needed for search mappings
+#include <linux/can.h>
 #include <sys/prctl.h>
+#include <sys/resource.h>
+#include <sys/socket.h>
+#include <sys/types.h>
 
 #include <seccomp.h>
 
-#include "seccomp.h"
+#include "seccomp_utils.h"
 #include "utils.h"
 
+#define sc_map_add(X) sc_map_add_kvp(#X, X)
+
+// libseccomp maximum per ARG_COUNT_MAX in src/arch.h
+#define SC_ARGS_MAXLENGTH	6
 #define SC_MAX_LINE_LENGTH	82	// 80 + '\n' + '\0'
 
 char *filter_profile_dir = "/var/lib/snapd/seccomp/profiles/";
+struct hsearch_data sc_map_htab;
+
+enum parse_ret {
+	PARSE_INVALID_SYSCALL = -2,
+	PARSE_ERROR = -1,
+	PARSE_OK = 0,
+};
 
 struct preprocess {
 	bool unrestricted;
 	bool complain;
 };
 
+/*
+ * arg_cmp contains items of type scmp_arg_cmp (from SCMP_CMP macro) and
+ * length is the number of items in arg_cmp that are active such that if
+ * length is '3' arg_cmp[0], arg_cmp[1] and arg_cmp[2] are used, when length
+ * is '1' only arg_cmp[0] and when length is '0', none are used.
+ */
+struct seccomp_args {
+	int syscall_nr;
+	unsigned int length;
+	struct scmp_arg_cmp arg_cmp[SC_ARGS_MAXLENGTH];
+};
+
+struct sc_map_entry {
+	ENTRY *e;
+	ENTRY *ep;
+	struct sc_map_entry *next;
+};
+
+struct sc_map_list {
+	struct sc_map_entry *list;
+	int count;
+};
+
+struct sc_map_list *sc_map_entries = NULL;
+
+/*
+ * Setup an hsearch map to map strings in the policy (eg, AF_UNIX) to
+ * scmp_datum_t values. Abstract away hsearch implementation behind sc_map_*
+ * functions in case we want to swap this out.
+ *
+ * sc_map_init()		- initialize the hash map via linked list of
+ * 				  of entries
+ * sc_map_add_kvp(key, value)	- create entry from key/value pair and add to
+ * 				  linked list
+ * sc_map_search(s)	- if found, return scmp_datum_t for key, else set errno
+ * sc_map_destroy()	- destroy the hash map and linked list
+ */
+scmp_datum_t sc_map_search(char *s)
+{
+	ENTRY e;
+	ENTRY *ep = NULL;
+	scmp_datum_t val = 0;
+	errno = 0;
+
+	e.key = s;
+	if (hsearch_r(e, FIND, &ep, &sc_map_htab) == 0)
+		die("hsearch_r failed");
+
+	if (ep != NULL) {
+		scmp_datum_t *p = ep->data;
+		val = *p;
+	} else
+		errno = EINVAL;
+
+	return val;
+}
+
+void sc_map_add_kvp(const char *key, scmp_datum_t value)
+{
+	struct sc_map_entry *node;
+	node = (struct sc_map_entry *)malloc(sizeof(struct sc_map_entry));
+	if (node == NULL)
+		die("Out of memory creating sc_map_entries");
+
+	node->e = (ENTRY *) malloc(sizeof(ENTRY));
+	if (node->e == NULL)
+		die("Out of memory creating ENTRY");
+
+	node->e->key = strdup(key);
+	if (node->e->key == NULL)
+		die("Out of memory creating e->key");
+
+	node->e->data = (void *)malloc(sizeof(scmp_datum_t));
```

Comment:
This is incorrect and shows the potential problem. This is allocating size of a pointer instead of the size of the type being used to store the data. But ignore this for now because I think I have a better fix to suggest in a higher-level comment.